### DISCUSSION TITLE
Solved: All devices and VMs should have unique names in Netbox.

### DISCUSSION THREAD

**Discussion by LHBL2003** [netbox-community/netbox#16267] [Q&A] [answered]
Devices are unique within a tenant. But is it also possible to say that all devices and VMs should have unique names?
Everything that doesn't need a unique name is simply not given a name in netbox. We use the tenants to group everything that belongs to a location. However, VMs and notebooks, for example, should have unique names within the company.

Hence the question whether there is a function / configuration field in which the name Device Name and VM Name must always be unique.

I've already told you where to do this. I just don't know how to do it.
![image](https://github.com/netbox-community/netbox/assets/46369917/143cc73a-663f-4f85-b18a-7ee3c6c9af1b)

```
{
    "dcim.device": [
        {
            "name": {
                "min_length": 5,
                "max_length": 30
            }
        }
    ]
}
```
https://docs.netbox.dev/en/stable/customization/custom-validation/

Addendum: I have just found something that looks like what I am looking for. However, I have no idea where to write such a validation class:
https://github.com/netbox-community/netbox/discussions/10122 Does anyone know where the file is where you write the CustomValidator Class? The documentation says that you can write a custom validation logic, but not where to write it.

**Comment by Mailstorm-ctrl** [comment]
You write your own custom validator (as a py file) and make it available to Netbox. Whether that be through a datasource or uploading directly to the server. For example, here's a custom validator I have that I've put at `/opt/netbox/validators`

```
from extras.validators import CustomValidator

class checkMK_TagValidator(CustomValidator):
        def validate(self, instance):
                if "CheckMK Monitored".upper() in [tag.name.upper() for tag in instance._m2m_values['tags']] and instance.primary_ip is None:
                        self.fail("CheckMK tag requires a primary IP on the device!", field='tags')
```

Then in the config I have the following for custom validators:

```
{
    "dcim.device": [
        "validators.tagValidate.checkMK_TagValidator"
    ]
}
```

**Reply by LHBL2003** [reply]
@Mailstorm-ctrl: Thanks for the info. I'll create a "validators" folder right away, because I don't have one.

> or uploading directly to the server.

What do you mean? Is there an area in the Netbox UI where you can do this?

**Comment by LHBL2003** [comment]
So that others don't have to go searching to find everything relevant for an example:

I have already managed to get it to report that a device with the same name already exists.

😃   ![image](https://github.com/netbox-community/netbox/assets/46369917/57a41c09-15f9-4820-ac23-c2e456d9edfa)

The file name must apparently be written in lower case, the name does not matter. 
`/opt/netbox/validators/custom.py`

```
from extras.validators import CustomValidator
from dcim.models import Device

class DeviceNameUnique(CustomValidator):
    def validate(self, device):
        existing_devices = Device.objects.filter(name__iexact=device.name).exists()
        print(existing_devices)
        if existing_devices:
            self.fail("Device name already in use.", field="name")
```

A restart of the Netbox service is not necessary if the "Custom validators" configuration is entered in the WebUI.
A restart is also not necessary if the .py file is changed.
![image](https://github.com/netbox-community/netbox/assets/46369917/c16479b4-2170-4fec-a11d-e546e723eba4)

```
{
    "dcim.device": [
        "validators.custom.DeviceNameUnique"
    ]
}
```
validators= Name of the folder (I hope)
custom = Name of the Python file
DeviceNameUnique = Name of the class in the .py file.
dcim.device = Can be found in a device in the top right-hand corner. (See picture)
![image](https://github.com/netbox-community/netbox/assets/46369917/cb28b67b-0e3b-402a-abcb-c7d55599c29a)

From now on, however, you must remember to copy your validators folder when updating. I don't think this is mentioned in the update instructions.

**Comment by LHBL2003** [comment]
I have no idea, but I know a lot about it. 😄 
Here is another example so that the name must also be unique for virtual machines.

![image](https://github.com/netbox-community/netbox/assets/46369917/db066c34-9dbf-4e5f-912c-cbba4f3b8c30)

**Python File**
```
from extras.validators import CustomValidator
from dcim.models import Device
from virtualization.models import VirtualMachine

class DeviceNameUnique(CustomValidator):
    def validate(self, device):
        existing_devices = Device.objects.filter(name__iexact=device.name).exists()
        print(existing_devices)
        if existing_devices:
            self.fail("Custom Validator: Device name already in use.", field="name")


class VirtualMachineNameUnique(CustomValidator):
    def validate(self, virtualmachine):
        existing_virtualmachines =VirtualMachine.objects.filter(name__iexact=virtualmachine.name).exists()
        print(existing_virtualmachines)
        if existing_virtualmachines:
            self.fail("Custom Validator: Virtual Machine name already in use.", field="name")
```

**Custom validators**
```
{
    "dcim.device": [
        "validators.custom.DeviceNameUnique"
    ],
    "virtualization.virtualmachine": [
        "validators.custom.VirtualMachineNameUnique"
    ]
}
```

The information to be provided must be derived from the path of these files and their classes. You will find the names from the script again and can adapt this to the topic of virtual machines.
https://github.com/netbox-community/netbox/blob/develop/netbox/dcim/models/devices.py
https://github.com/netbox-community/netbox/blob/develop/netbox/virtualization/models/virtualmachines.py

**Comment by LHBL2003** [comment]
Oh man, it doesn't work like that. 
It now also checks this when I change the status in a device.
So with every change it says that this device already exists. Does anyone know how to change the classes so that it doesn't look at itself?

**Reply by candlerb** [reply]
> Does anyone know how to change the classes so that it doesn't look at itself?

You could try `filter(name__iexact=device.name, id__n=device.pk)`

**Reply by LHBL2003** [reply]
Do I still have to import any classes?

![image](https://github.com/netbox-community/netbox/assets/46369917/f50bff35-1c84-4364-a672-dedfeb1ba521)

```
from extras.validators import CustomValidator
from dcim.models import Device
from virtualization.models import VirtualMachine

# Der Device Name muss im gesammten Netbox System eindeutig sein.
class DeviceNameUnique(CustomValidator):
    def validate(self, device):
        existing_devices = Device.objects.filter(name__iexact=device.name, id__n=device.pk).exists()
        print(existing_devices)
        if existing_devices:
            self.fail("Custom Validator: Device name already in use.", field="name")
```

**Reply by candlerb** [reply]
You're right, `__n` is a REST API query filter, not a Django ORM.

Another attempt: `filter(name__iexact=device.name).exclude(id=device.pk)`

**Comment by LHBL2003** [comment, marked as answer]
**My thanks go to:**
@domainzero --> Creator of the base script on which I have built. 
@Mailstorm-ctrl --> Who helped me reach my goal.
@candlerb --> Who helped me reach my goal.

**Final version**
You can find everything you need to understand it in the entries above.

**Description**
This version allows the name of a virtual machine and of a device to appear only once in the entire Netbox system. Changing the VM/device also works in this version. (The comments in the script are in German. Increases the Google hits for German ;)

**WebUI: Configuration History --> Custom vakidators**

```
{
    "__comment": "Der Device / VM Name soll im gesamten Netbox System eindeutig sein. (Github Netbox discussions 16267)",
    "dcim.device": [
        "validators.custom.DeviceNameUnique"
    ],
    "virtualization.virtualmachine": [
        "validators.custom.VirtualMachineNameUnique"
    ]
}
```

**/opt/netbox/validators/custom.py**
```
# Dies sind selbst erstellte Validierungsprüfungen auf Python Basis
# Merh Informationen hierzu findest du unter: https://github.com/netbox-community/netbox/discussions/16267

from extras.validators import CustomValidator
from dcim.models import Device
from virtualization.models import VirtualMachine

# Der Device Name muss im gesammten Netbox System eindeutig sein.
class DeviceNameUnique(CustomValidator):
    def validate(self, device):
        existing_devices = Device.objects.filter(name__iexact=device.name).exclude(id=device.pk).exists()
        print(existing_devices)
        if existing_devices:
            self.fail("Custom Validator: Device name already in use.", field="name")

# Der VM Name muss innerhalb des gesammten Netbox Systems eindeutig sein.
class VirtualMachineNameUnique(CustomValidator):
    def validate(self, virtualmachine):
        existing_virtualmachines =VirtualMachine.objects.filter(name__iexact=virtualmachine.name).exclude(id=virtualmachine.pk).exists()
        print(existing_virtualmachines)
        if existing_virtualmachines:
            self.fail("Custom Validator: Virtual Machine name already in use.", field="name")
```

**Reply by candlerb** [reply]
Great. I think you might be able to simplify it to work with any model that has a "name" field:

```
from extras.validators import CustomValidator

class NameUnique(CustomValidator):
    def validate(self, obj):
        if obj.__class__.objects.filter(name__iexact=obj.name).exclude(id=obj.pk).exists():
            self.fail("Custom Validator: Name already in use.", field="name")
```

**Reply by LHBL2003** [reply]
I'll keep that in mind, but that would mean that everything that has a name in Netbox can only exist once. I'll have to think about whether that's a good idea. Because then a location would certainly also have to be unique.

**Reply by candlerb** [reply]
You would only apply that validator to those models which you want unique names. In your case you only want unique names for devices and VMs, so it would be:

```
{
    "__comment": "Der Device / VM Name soll im gesamten Netbox System eindeutig sein. (Github Netbox discussions 16267)",
    "dcim.device": [
        "validators.custom.NameUnique"
    ],
    "virtualization.virtualmachine": [
        "validators.custom.NameUnique"
    ]
}
```

It just saves a bit of code duplication.

**Reply by LHBL2003** [reply]
Oh, that's great too. THX

I've just realised that the check also works if you don't specify a device name. That's why I added "if obj.name and" so that the check is not carried out if the Name field remains empty. This is because devices without a name should still be able to be created.

**Generalised and Still allows empty device names.**
```
from extras.validators import CustomValidator

class NameUnique(CustomValidator):
    def validate(self, obj):
        if obj.name and obj.__class__.objects.filter(name__iexact=obj.name).exclude(id=obj.pk).exists():
            self.fail("Custom Validator: Name already in use.", field="name")
```

For all others: I now believe that I have to restart Netbox. Because I sometimes suspect that my change might not work otherwise.
`systemctl restart netbox`

**Comment by LHBL2003** [comment]
Does anyone happen to know how to check a customer field for uniqueness?

Customer Fields Name: DCIM_Device_InventoryNo
Type: Text
